Refactor ERC-20 Interactions in Driver

nlordell opened the pull request:
This PR just proposes a small refactor in the `driver` to make code that uses ERC-20s easier. I also flattened the `Spender` type, since having the token tied to the spender felt a bit awkward.

It introduces a `blockchain::token::Erc20` type for fetching ERC-20 token information.

### Test Plan

Rust CI - tests continue to pass.

MartinquaXD commented on crates/driver/src/infra/blockchain/token.rs:
```diff
@@ -0,0 +1,85 @@
+    /// Fetch the ERC20 allowance for the spender. See the allowance method in
+    /// EIP-20.
+    ///
+    /// https://eips.ethereum.org/EIPS/eip-20#methods
```
I guess the links above the function implementations could either be dropped or modified to link to the actual function on the website.

nlordell replied:
They point to the `methods` section (Which includes all ERC-20 contract method documentation). It was mostly inherited from the `allowance` function documentation.

MartinquaXD replied:
Yes. But it's also possible to link to the exact method that is getting implemented.

nlordell replied:
🤯

nlordell replied:
Fixed.

MartinquaXD commented on crates/driver/src/infra/blockchain/token.rs:
```diff
@@ -0,0 +1,85 @@
+///
+/// https://eips.ethereum.org/EIPS/eip-20
+pub struct Erc20 {
+    contract: contracts::ERC20,
```
Nice component. 👍

fleupold reviewed (approved):
Looks good, but I don't see where you limit the allowance to the requested one in the code (was expecting that given the PR title)

nlordell commented:
> but I don't see where you limit the allowance to the requested one in the code (was expecting that given the PR title)

Yes, sorry - I forgot to change the title. I was looking through the `driver` code and noticed we do `approve(spender, type(uint256).max)`. For some reason, I (mis)remembered that we had decided to change this behaviour in the `solver` crate - but it turns out that we didn't actually change it - so I ended up leaving the refactor I had and removed the change to set the exact approval that the solvers requested.

I updated the PR title.
